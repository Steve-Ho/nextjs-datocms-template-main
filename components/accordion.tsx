import {useState, useContext, useCallback, useMemo, createContext} from 'react';
import type * as React from 'react';
import {isNil, reject, equals} from 'remeda';
import {ChevronUpIcon, ChevronDownIcon} from '@heroicons/react/solid';

type ContainerContext = {
	toggleActivePanel?: ((label: string) => void);
	activePanels: string[];
};

type ExpandMode = 'multiple' | 'single';

type ContainerProperties = {
	expandMode?: ExpandMode;
	children: JSX.Element[];
};

type PanelProperties = {
	children: JSX.Element | JSX.Element[];
	label: string;
	icon?: React.ElementType;
};

const containerContext = createContext<ContainerContext>({
	activePanels: [],
});

const Container = ({children, expandMode = 'single'}: ContainerProperties): JSX.Element => {
	const [activePanels, setActivePanels] = useState<string[]>([]);

	const toggleActivePanel = useCallback((label: string) => {
		if (expandMode === 'single') {
			setActivePanels(
				activePanels.includes(label)
					? []
					: [label],
			);
		}

		if (expandMode === 'multiple') {
			setActivePanels(
				activePanels.includes(label)
					? reject(activePanels, equals(label))
					: [...activePanels, label],
			);
		}
	}, [activePanels, expandMode]);

	const context = useMemo(() => ({
		toggleActivePanel,
		activePanels,
	}), [toggleActivePanel, activePanels]);

	return (
		<section className="space-y-3">
			<containerContext.Provider value={context}>
				{children}
			</containerContext.Provider>
		</section>
	);
};

const generatePanelButtonClasses = (isActive: boolean): string => `
	flex
	w-full
	items-center
	space-x-2
	flex-grow
	text-left
	subheading
	py-2
	border-b-2
	${isActive ? 'border-true-gray-400' : 'border-true-gray-200'}
	${isActive ? '' : 'text-true-gray-400'}
`;

const Panel = ({children, label, icon: Icon}: PanelProperties): JSX.Element => {
	const {toggleActivePanel, activePanels} = useContext(containerContext);

	const handleTogglePanel = useCallback(() => {
		toggleActivePanel?.(label);
	}, [toggleActivePanel, label]);

	const isActive = useMemo(() => activePanels.includes(label), [activePanels, label]);

	const classes = useMemo(() => ({
		button: generatePanelButtonClasses(isActive),
	}), [isActive]);

	return (
		<article>
			<header>
				<button key={label} className={classes.button} type="button" onClick={handleTogglePanel}>
					{!isNil(Icon) && <Icon className="w-5"/>}
					<span>{label}</span>
					{isActive && <ChevronUpIcon className="!ml-auto w-5"/>}
					{!isActive && <ChevronDownIcon className="!ml-auto w-5"/>}
				</button>
			</header>
			<div className="mt-4">
				{isActive && children}
			</div>
		</article>
	);
};

const Accordion = {
	Container,
	Panel,
};

export default Accordion;

export type {
	ContainerContext,
	ExpandMode,
	ContainerProperties as ContainerProps,
	PanelProperties as PanelProps,
};
