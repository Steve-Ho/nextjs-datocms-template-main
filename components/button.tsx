import {forwardRef, useMemo} from 'react';
import type * as React from 'react';
import {isNil} from 'remeda';
import {collapse} from '@growthops/ext-ts';

type Variant =
	| 'primary'
	| 'secondary'
	| 'tertiary';

type Size =
	| 'large'
	| 'regular'
	| 'small';

type IconAlignment = 'left' | 'right';

type Icon = {
	svg: React.ElementType;
	alignment?: IconAlignment;
	className?: string;
};

type BaseButtonProperties = {
	label: string;
	className?: string;
	variant?: Variant;
	size?: Size;
	icon?: Icon;
};

type SemanticButtonProperties = BaseButtonProperties & React.ComponentPropsWithoutRef<'button'>;

type LinkButtonProperties = BaseButtonProperties & React.ComponentPropsWithoutRef<'a'>;

const baseClasses = 'inline-flex items-center rounded !leading-none';

const variantClasses: Record<Variant, string> = {
	primary: 'bg-true-gray-600 text-white',
	secondary: 'bg-true-gray-300 text-true-gray-600',
	tertiary: 'border border-true-gray-300 text-true-gray-700',
};

const sizeClasses: Record<Size, string> = {
	large: 'px-8 py-5 heading-four space-x-4',
	regular: 'px-6 py-4 heading-six space-x-3',
	small: 'px-5 py-3 text-small space-x-2',
};

const iconClasses: Record<Size, string> = {
	large: 'w-7',
	regular: 'w-5',
	small: 'w-4',
};

const generateIcon = (
	icon: Icon | undefined,
	alignment: IconAlignment,
	classes: string,
	alignmentAdjustmentClasses: string,
): JSX.Element | undefined => {
	if (!isNil(icon) && (icon.alignment ?? 'left') === alignment) {
		return <icon.svg className={collapse([classes, alignmentAdjustmentClasses])}/>;
	}
};

const useClasses = (
	variant: Variant, size: Size, className: string | undefined, icon: Icon | undefined,
): Record<string, string> => useMemo(
	() => ({
		root: collapse([baseClasses, variantClasses[variant], sizeClasses[size], className ?? '']),
		icon: collapse([iconClasses[size], icon?.className ?? '']),
	}),
	[variant, size, className, icon?.className],
);

const useIcons = (icon: Icon | undefined, classes: string): Record<string, JSX.Element | undefined> => useMemo(() => ({
	left: generateIcon(icon, 'left', classes, '-ml-2'), // -ml-* here to visually weight the icon appropriately.
	right: generateIcon(icon, 'right', classes, '!-mr-2'), // !-mr-* here to visually weight the icon appropriately.
}), [icon, classes]);

const Semantic = ({
	label, variant = 'primary', size = 'regular', className, icon, ...intrinsicButtonProperties
}: SemanticButtonProperties): JSX.Element => {
	const classes = useClasses(variant, size, className, icon);
	const icons = useIcons(icon, classes.icon);

	return (
		<button className={classes.root} {...intrinsicButtonProperties} type="button">
			{icons.left}
			<span>{label}</span>
			{icons.right}
		</button>
	);
};

const Link = forwardRef<HTMLAnchorElement, LinkButtonProperties>(({
	label, variant = 'primary', size = 'regular', className, icon, ...intrinsicAnchorProperties
}: LinkButtonProperties, reference): JSX.Element => {
	const classes = useClasses(variant, size, className, icon);
	const icons = useIcons(icon, classes.icon);

	return (
		<a ref={reference} className={classes.root} {...intrinsicAnchorProperties} rel="noreferrer">
			{icons.left}
			<span>{label}</span>
			{icons.right}
		</a>
	);
});

Link.displayName = 'Link';

const Button = {
	Semantic,
	Link,
};

export default Button;

export type {
	Variant,
	Size,
	IconAlignment,
	Icon,
	BaseButtonProperties as BaseButtonProps,
	SemanticButtonProperties as SemanticButtonProps,
	LinkButtonProperties as LinkButtonProps,
};
