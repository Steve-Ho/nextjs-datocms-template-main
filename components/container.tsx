import {useMemo} from 'react';
import {collapse} from '@growthops/ext-ts';

type ContainerProperties = {
	children: Array<JSX.Element | boolean> | JSX.Element | boolean;
	className?: string;
};

const Container = ({className = '', children}: ContainerProperties): JSX.Element => {
	const classes = useMemo(() => ({
		root: collapse(['container', className]),
	}), [className]);

	return (
		<div className={classes.root}>
			{children}
		</div>
	);
};

export default Container;

export type {
	ContainerProperties as ContainerProps,
};
