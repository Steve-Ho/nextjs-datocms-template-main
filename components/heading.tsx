import {collapse} from '@growthops/ext-ts';

type Variant =
	| 'heading-five'
	| 'heading-four'
	| 'heading-one'
	| 'heading-six'
	| 'heading-three'
	| 'heading-two'
	| 'headline'
	| 'subheading';

type SemanticElement =
	| 'h1'
	| 'h2'
	| 'h3'
	| 'h4'
	| 'h5'
	| 'h6'
	| 'p'
	| 'span';

type HeadingProperties = {
	variant: Variant;
	label: string;
	element?: SemanticElement;
	className?: string;
};

const Heading = ({variant, label, element = 'p', className = ''}: HeadingProperties): JSX.Element => {
	const sharedProperties = {
		className: collapse([variant, className]),
	};

	switch (element) {
		case 'h1': return <h1 {...sharedProperties}>{label}</h1>;
		case 'h2': return <h2 {...sharedProperties}>{label}</h2>;
		case 'h3': return <h3 {...sharedProperties}>{label}</h3>;
		case 'h4': return <h4 {...sharedProperties}>{label}</h4>;
		case 'h5': return <h5 {...sharedProperties}>{label}</h5>;
		case 'h6': return <h6 {...sharedProperties}>{label}</h6>;
		case 'p': return <p {...sharedProperties}>{label}</p>;
		case 'span': return <span {...sharedProperties}>{label}</span>;
	}
};

export default Heading;

export type {
	Variant,
	SemanticElement,
	HeadingProperties as HeadingProps,
};
