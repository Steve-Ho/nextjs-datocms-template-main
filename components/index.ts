export {default as Accordion} from './accordion';
export {default as Button} from './button';
export {default as Card} from './card';
export {default as Checkbox} from './checkbox';
export {default as Container} from './container';
export {default as Heading} from './heading';
export {default as Input} from './input';
export {default as PreviewMode} from './preview-mode';
export {default as ProgressBar} from './progress-bar';
export {default as Radio} from './radio';
export {default as Region} from './region';
export {default as Select} from './select';
export {default as Tab} from './tab';
export {default as Text} from './text';
export {default as Textarea} from './textarea';
