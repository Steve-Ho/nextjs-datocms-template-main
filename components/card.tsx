import {useMemo} from 'react';
import type {ResponsiveImageType} from 'react-datocms';
import {Image} from 'react-datocms';
import {isNil} from 'remeda';
import {collapse} from '@growthops/ext-ts';
import {Heading, Text} from '@app/components';

type CardProperties = {
	image: ResponsiveImageType;
	heading: string;
	content: string;
	isFramed?: boolean;
	hasAutoHeight?: boolean;
	meta?: JSX.Element;
};

const generateRootClasses = (isFramed: boolean): string => `
	overflow-hidden
	rounded
	h-full
	flex
	flex-col
	${isFramed ? 'bg-white shadow-lg' : ''}
`;

const generateBodyClasses = (isFramed: boolean): string => `
	flex
	flex-col
	flex-grow
	${isFramed ? 'px-4 pb-3' : ''}
`;

const generateMetaClasses = (hasAutoHeight: boolean): string => `
	pt-6
	${hasAutoHeight ? 'mt-auto' : ''}
`;

const generateImageClasses = (isFramed: boolean): string => `
	w-full
	${isFramed ? '' : 'rounded shadow-lg'}
`;

const Card = ({image, heading, content, meta, isFramed = false, hasAutoHeight = false}: CardProperties): JSX.Element => {
	const classes = useMemo(() => ({
		root: collapse(generateRootClasses(isFramed)),
		image: collapse(generateImageClasses(isFramed)),
		body: collapse(generateBodyClasses(isFramed)),
		heading: 'mt-4',
		content: 'mt-2',
		meta: collapse(generateMetaClasses(hasAutoHeight)),
	}), [isFramed, hasAutoHeight]);

	return (
		<article className={classes.root}>
			<Image className={classes.image} data={image}/>
			<div className={classes.body}>
				<Heading className={classes.heading} label={heading} variant="heading-four"/>
				<Text className={classes.content} variant="text-regular">{content}</Text>
				{!isNil(meta) && (
					<div className={classes.meta}>
						{meta}
					</div>
				)}
			</div>
		</article>
	);
};

export default Card;

export type {
	CardProperties as CardProps,
};
