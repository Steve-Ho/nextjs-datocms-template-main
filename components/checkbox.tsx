import {forwardRef} from 'react';
import type * as React from 'react';
import {collapse} from '@growthops/ext-ts';
import {FormField} from '@app/components/utility';
import type {FormFieldInputProps} from '@app/components/utility/form-field';
import {CheckIcon} from '@heroicons/react/solid';

type CheckboxProperties = FormFieldInputProps &
	React.ComponentPropsWithoutRef<'input'>;

const checkboxWrapperClasses = collapse(`
	flex
	items-center
	mr-4
	py-1
	mt-3
`);

const checkboxClasses = collapse(`
	opacity-0
	absolute
	h-8
	w-8
`);

const checkboxSVGWrapperClasses = collapse(`
	bg-white
	border-2
	rounded-sm
	border-gray-400
	w-5
	h-5
	flex
	flex-shrink-0
	justify-center
	items-center
	mr-2
	focus-within:border-gray-500
`);

const checkboxSVGClasses = collapse(`
	fill-current
	hidden
	w-3
	h-3
	text-gray-600
	pointer-events-none
	bg-gray-300
	rounded-sm
`);

const Checkbox = forwardRef<HTMLInputElement, CheckboxProperties>(({
	className, id, label, error, helpText, ...intrinsicCheckboxProperties
}: CheckboxProperties, reference) => (
	<FormField
		{...{className, id, label, helpText, error}}
		isRequired={intrinsicCheckboxProperties.required}
		isDisabled={intrinsicCheckboxProperties.disabled}
	>
		{(label, baseClasses) => (
			<div className={checkboxWrapperClasses}>
				<input
					ref={reference}
					id={id}
					type="checkbox"
					className={collapse([baseClasses, checkboxClasses])}
					{...intrinsicCheckboxProperties}
				/>
				<div className={checkboxSVGWrapperClasses}>
					<CheckIcon className={checkboxSVGClasses}/>
				</div>
				{label}
			</div>
		)}
	</FormField>
));

Checkbox.displayName = 'Checkbox';

export default Checkbox;

export type {CheckboxProperties as CheckboxProps};
