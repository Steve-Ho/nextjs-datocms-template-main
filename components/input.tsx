import {forwardRef} from 'react';
import * as React from 'react';
import {collapse} from '@growthops/ext-ts';
import {FormField} from '@app/components/utility';
import type {FormFieldInputProps} from '@app/components/utility/form-field';

type InputProperties = FormFieldInputProps & React.ComponentPropsWithoutRef<'input'>;

const Input = forwardRef<HTMLInputElement, InputProperties>(({
	className, id, label, error, helpText, ...intrinsicInputProperties
}: InputProperties, reference) => (
	<FormField
		{...{className, id, label, helpText, error}}
		isRequired={intrinsicInputProperties.required}
		isDisabled={intrinsicInputProperties.disabled}
	>
		{(label, baseClasses) => (
			<React.Fragment>
				{label}
				<input ref={reference} id={id} className={collapse(baseClasses)} {...intrinsicInputProperties}/>
			</React.Fragment>
		)}
	</FormField>
));

Input.displayName = 'Input';

export default Input;

export type {
	InputProperties as InputProps,
};
