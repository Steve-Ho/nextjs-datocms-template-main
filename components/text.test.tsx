import {expect, test} from '@jest/globals';
import {create} from 'react-test-renderer';
import {Text} from '@app/components';
import type {Variant, SemanticElement} from './text';

const variants: Variant[] = ['text-lead', 'text-regular', 'text-small'];
const elements: SemanticElement[] = ['p', 'span'];

test('snapshots', () => {
	for (const variant of variants) {
		for (const element of elements) {
			const result = create(
				<Text variant={variant} element={element}>
					Hello World
				</Text>,
			).toJSON();

			expect(result).toMatchSnapshot();
		}
	}
});
