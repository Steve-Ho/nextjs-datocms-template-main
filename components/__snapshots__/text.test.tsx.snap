// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots 1`] = `
<p
  className="text-lead"
>
  Hello World
</p>
`;

exports[`snapshots 2`] = `
<span
  className="text-lead"
>
  Hello World
</span>
`;

exports[`snapshots 3`] = `
<p
  className="text-regular"
>
  Hello World
</p>
`;

exports[`snapshots 4`] = `
<span
  className="text-regular"
>
  Hello World
</span>
`;

exports[`snapshots 5`] = `
<p
  className="text-small"
>
  Hello World
</p>
`;

exports[`snapshots 6`] = `
<span
  className="text-small"
>
  Hello World
</span>
`;
